analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -286,13 +282,25 @@ public void Branching_LearnsObjectConstraint_Binary_UndefinedInOtherBranch_CS(st
                 .And.ContainSingle(x => x != null && x.HasConstraint(ObjectConstraint.NotNull));
         }
 
+        [DataTestMethod]
+        [DataRow("arg == true", true)]
+        [DataRow("arg == false", false)]
+        [DataRow("true == arg", true)]
+        [DataRow("false == arg", false)]
+        public void Branching_LearnsObjectConstraint_Binary_UndefinedInOtherBranch_NullableBool_CS(string expression, bool expected)
+        {
+            var expectedBoolConstraint = BoolConstraint.From(expected);
+            var validator = CreateIfElseEndValidatorCS(expression, OperationKind.Binary, "bool?");
+            validator.ValidateTag("If", x => x.Should().HaveOnlyConstraints(ObjectConstraint.NotNull, expectedBoolConstraint));
```

Comment:
This might be out of scope for this PR because it concerns all tests in this file, but I think it would be better to not check for the absence of constraints not relevant to the test and only focus on one constraint type per test. So one test for each:

- "arg == true" learns NotNull
- "arg == true" does not learn Null
- "arg == true" learns True
- "arg == true" does not learn False

Benefits:

- Implementing more constraints doesn't break existing tests.
- Analyzing tests will be easier.

Reply:
If the planned number of constraints is limited, this is fine. But the approach does not scale well for big numbers.

Reply:
There will be practically just one more constraint built by the engine for integer numbers. Everything else is rule-specific. And once we add that, we'll need to see it's effect (or undesired effect) on all the various cases we have. So we're fine.

Reply:
..in general, there's not much more we can learn about `true` or `false` values (and similar) anyway.